runtime/src/main/scala/akka/grpc/internal/ServerReflectionImpl.scala

```diff
@@ -0,0 +1,108 @@
+        case In.ListServices(_) =>
+          val list = services.map(s => ServiceResponse(s))
+          Out.ListServicesResponse(ListServiceResponse(list))
+      }
+      // TODO Validate assumptions here
+      ServerReflectionResponse(req.host, Some(req), response)
+    })
+  }
+}
+object ServerReflectionImpl {
+  import scala.collection.JavaConverters._
+
+  def apply(fileDescriptors: Seq[FileDescriptor], services: List[String]): ServerReflectionImpl =
+    new ServerReflectionImpl(
+      (AdditionalDescriptors ++ fileDescriptors).map(fd => fd.getName -> fd).toMap,
+      ServerReflection.name +: services)
+
+  val AdditionalDescriptors =
+    flattenDependencies(ReflectionProto.javaDescriptor).distinct
+
+  private def flattenDependencies(descriptor: FileDescriptor): List[FileDescriptor] = {
+    descriptor :: descriptor.getDependencies.asScala.toList.flatMap(flattenDependencies)
+  }
+
+  def splitNext(name: String): (String, String) = {
+    val dot = name.indexOf('.')
+    if (dot >= 0) {
+      (name.substring(0, dot), name.substring(dot + 1))
+    } else {
+      (name, "")
+    }
+  }
+
+  def containsSymbol(symbol: String, fileDesc: FileDescriptor): Boolean =
+    (symbol.startsWith(fileDesc.getPackage)) && // Ensure package match first
+    (splitNext(if (fileDesc.getPackage.isEmpty) symbol else symbol.drop(fileDesc.getPackage.length + 1)) match {
+      case ("", "")            => false
+      case (typeOrService, "") =>
+        //fileDesc.findEnumTypeByName(typeOrService) != null || // TODO investigate if this is expected
+        fileDesc.findMessageTypeByName(typeOrService) != null ||
+        fileDesc.findServiceByName(typeOrService) != null
+      case (service, method) =>
+        Option(fileDesc.findServiceByName(service)).exists(_.findMethodByName(method) != null)
+    })
+
+  def findFileDescForSymbol(symbol: String, fileDescriptors: Map[String, FileDescriptor]): Option[FileDescriptor] =
+    fileDescriptors.values.collectFirst {
+      case fileDesc if containsSymbol(symbol, fileDesc) => fileDesc
+    }
+
+  def containsExtension(container: String, number: Int, fileDesc: FileDescriptor): Boolean =
+    fileDesc.getExtensions.iterator.asScala.exists(ext =>
+      container == ext.getContainingType.getFullName && number == ext.getNumber)
+
+  def findFileDescForExtension(
+      container: String,
+      number: Int,
+      fileDescriptors: Map[String, FileDescriptor]): Option[FileDescriptor] =
+    fileDescriptors.values.collectFirst {
+      case fileDesc if containsExtension(container, number, fileDesc) => fileDesc
+    }
+
+  def findExtensionNumbersForContainingType(
+      container: String,
+      fileDescriptors: Map[String, FileDescriptor]): List[Int] =
+    (for {
+      fileDesc <- fileDescriptors.values.iterator
+      extension <- fileDesc.getExtensions.iterator.asScala
+      if extension.getFullName == container
+    } yield extension.getNumber).toList
+}
```

Comment:
Some unit tests for these contains/find methods would be nice, both for verifying that they do what they should and to document how they are expected to behave.